internal/service/alarms/internal/alarms_server.go

```diff
@@ -363,11 +364,43 @@ func (a *AlarmsServer) GetProbableCause(ctx context.Context, request api.GetProb
 	return nil, fmt.Errorf("not implemented")
 }
 
+// AmNotification handles an API request coming from AlertManager with CaaS alerts. This api is used internally.
+// Note: the errors returned can also be view under alertmanager pod logs but also logging here for convenience
 func (a *AlarmsServer) AmNotification(ctx context.Context, request api.AmNotificationRequestObject) (api.AmNotificationResponseObject, error) {
-	// TODO: Implement the logic to handle the AM notification
-	slog.Debug("Received AM notification", "groupLabels", request.Body.GroupLabels)
-	for _, alert := range request.Body.Alerts {
-		slog.Debug("Alert", "fingerprint", alert.Fingerprint, "startsAt", alert.StartsAt, "status", alert.Status)
+	// Audit the table with full list of alerts in the current payload. If missing set them to resolve
+	if err := a.AlarmsRepository.ResolveNotificationIfNotInCurrent(ctx, request.Body); err != nil {
+		msg := "failed to resolve notification that are not present"
+		slog.Error(msg, "error", err)
+		return nil, fmt.Errorf("%s: %w", msg, err)
+	}
+
+	// Get the definition data based on current set of Alert names and managed cluster ID
+	alarmDefinitions, err := a.AlarmsRepository.GetAlarmDefinitions(ctx, request.Body, a.ClusterServer.ClusterIDToResourceTypeID)
+	if err != nil {
+		msg := "failed to get AlarmDefinitions"
+		slog.Error(msg, "error", err)
+		return nil, fmt.Errorf("%s: %w", msg, err)
+	}
+
+	// Combine possible definitions with events
+	aerModels := alertmanager.ConvertAmToAlarmEventRecordModels(request.Body, alarmDefinitions, a.ClusterServer.ClusterIDToResourceTypeID)
+
+	// Insert and update AlarmEventRecord
+	if err := a.AlarmsRepository.UpsertAlarmEventRecord(ctx, aerModels); err != nil {
```

Comment:
I don't see why we need to upsert the current list. Above, you resolve the ones that are not in the list. That is good. Then, something that I'm missing is resolving the ones that explicitly come with firing status resolved. They are gonna be gone anyway in the next notification. But we don't know when this will happen. The `group_interval` states when a change is reported (new alert added to the list or one removed/resolved) but if nothing happens, this notification won't arrive when this timeout expires.

But anyway, the alerts are static, once they are reported, they are not gonna change. Only thing that changes is the firing status. You can perfectly identify with a query in the DB which alerts do not have a record in the DB. These are completely new, and you could simply add them. No need to process all of them

Reply:
Since dropping the archive table this may not be relevant anymore since we need to upsert the resolved data now.